src/components/PopupWithForm.js

```diff
@@ -39,11 +43,43 @@ export default class PopupWithForm extends Popup {
     return attributeNames;
   }
 
+  _changeSubmitButtonText(text) {
+    this._submitButton.textContent = text;
+  }
+
+  //takes a promise 'fetchRequest', and runs a saving animation on the submit button, which runs a minimum time, and doesn't complete until fetchRequest completes
+  async applySavingAnimation(fetchRequest) {
+    function delay(ms) {
+      return new Promise((resolve) => setTimeout(resolve, ms));
+    }
+    this._submitButton.classList.add("modal__submit-button_hover_disabled");
+    this._submitButton.classList.add("modal__submit-button_animation_active");
+    const [fetchResult, _] = await Promise.all([fetchRequest, delay(1200)]);
+    this._submitButton.classList.remove(
+      "modal__submit-button_animation_active"
+    );
+    this._submitButton.classList.add("modal__submit-button_saved");
+    await delay(180);
+    this._submitButton.classList.remove("modal__submit-button_hover_disabled");
+    return fetchResult;
+  }
+
+  open() {
+    if (this._submitButton.textContent !== this._defaultSubmitButtonText) {
+      this._changeSubmitButtonText(this._defaultSubmitButtonText);
+    }
+    if (this._submitButton.classList.contains("modal__submit-button_saved")) {
+      this._submitButton.classList.remove("modal__submit-button_saved");
+    }
+    super.open();
+  }
+
   setEventListeners() {
     //add the submit event handler to the form
-    this._formElement.addEventListener("submit", (e) => {
+    this._formElement.addEventListener("submit", async (e) => {
+      //this._changeSubmitButtonText("Saving...");
       e.preventDefault();
-      this._handleFormSubmit(this._getInputValues());
+      await this._handleFormSubmit(this._getInputValues());
```

Comment:
Good job that you have `await ` here to close the modal only after the server responds